File: problem_solving/dp/py/bs_122_buy_sell_stock_2_MULTIPLE_BS.py
'''
#https://leetcode.com/problems/best-time-to-buy-and-sell-stock-ii/

repeat 1 & 2 using while loop
    1) buy at the lowest .. using while
    2) sell at highest using while

'''
def buy_sell_stock(prices: list[int]):
    N=len(prices)
    i = 0
    buy = 0
    profit = 0
    while i < N-1:
        while i < N-1 and prices[i] >= prices[i+1]:
            i += 1
        buy = prices[i]
        #print("buy: " + str(buy))
        #sell
        while i < N-1 and prices[i] <= prices[i+1]:
            i += 1
        sell = prices[i]
        #print("sell: " + str(sell))
        profit += sell - buy
        #print("profit: " + str(profit))
    return profit


def buy_sell_stock_effective_buy_price(prices: list[int]):
    N=len(prices)
    profit = 0
    effective_buy_price = prices[0]

    for i in range(1, N):
        profit = max(profit, prices[i] - effective_buy_price)
        effective_buy_price = min(effective_buy_price, prices[i] - profit)
    return profit




prices = [7,6,4,1,5,8,3,6,4]
profit = buy_sell_stock(prices)
print("profit: " + str(profit))
profit = buy_sell_stock_effective_buy_price(prices)
print("profit-effective: " + str(profit))

#10


prices = [7,1,5,3,6,4]
#7
profit = buy_sell_stock(prices)
print("profit: " + str(profit))
profit = buy_sell_stock_effective_buy_price(prices)
print("profit-effective: " + str(profit))
#-------------------------------------

prices = [7,6,4,3,1]
#0
profit = buy_sell_stock(prices)
print("profit: " + str(profit))
profit = buy_sell_stock_effective_buy_price(prices)
print("profit-effective: " + str(profit))


prices = [1,2,3,4,5]
#4
profit = buy_sell_stock(prices)
print("profit: " + str(profit))
profit = buy_sell_stock_effective_buy_price(prices)
print("profit-effective: " + str(profit))


prices = [3,3,5,0,0,3,1,4]
#Output: 6
profit = buy_sell_stock(prices)
print("profit: " + str(profit))
profit = buy_sell_stock_effective_buy_price(prices)
print("profit-effective: " + str(profit))

File: problem_solving/dp/cpp/bs_122_buy_sell_stock_2_MULTIPLE_BS.cpp
/*
https://leetcode.com/problems/best-time-to-buy-and-sell-stock-ii/description/
*/

#include<vector>
#include<iostream>

int max_profit_multiple_tx(const std::vector<int>& prices) {
    int profit = 0;
    int n = prices.size();
    int buy = 0;
    int i = 0;
    while(i < n-1) {
        while(i < n-1 && prices[i] >= prices[i+1]) {
            i += 1;
        }
        buy = prices[i];

        while(i < n-1 && prices[i] <= prices[i+1]) {
            i += 1;
        }
        if(i < n && prices[i] > buy) {
            profit += prices[i] - buy;
        }
    }
    return profit;
}


int max_profit_multiple_tx_simplified(const std::vector<int>& prices) {
    int profit = 0;
    for(int i=1; i < prices.size(); ++i) {
        if(prices[i] > prices[i-1]) {
            profit += prices[i] - prices[i-1];
        }
    }
    return profit;
}


int buy_sell_stock_effective_buy_price(const std::vector<int>& prices) {
    int profit = 0;
    //#todo
    return profit;
}

int main() {
    std::vector<int> prices = {7,1,5,3,6,4};
    //Output: 7
    int profit = max_profit_multiple_tx(prices);
    std::cout << "result: " << profit << std::endl;
    profit = max_profit_multiple_tx_simplified(prices);
    std::cout << "result-simplify: " << profit << std::endl;

    prices = {1,2,3,4,5};
    //Output: 4
    profit = max_profit_multiple_tx(prices);
    std::cout << "result: " << profit << std::endl;
    profit = max_profit_multiple_tx_simplified(prices);
    std::cout << "result-simplify: " << profit << std::endl;

    prices = {7,6,4,3,1};
    //Output: 0}
    profit = max_profit_multiple_tx(prices);
    std::cout << "result: " << profit << std::endl;
    profit = max_profit_multiple_tx_simplified(prices);
    std::cout << "result-simplify: " << profit << std::endl;
}